File: apps/trading_kernel/lib/trading_kernel/backtest.ex
defmodule TradingKernel.Backtest do

  alias TradingSystem.Accounts
  alias TradingSystem.Markets
  alias TradingKernel.Common


  def run(symbol, user) do
    {position, data, config} = init_data(symbol, user)
    trading(position, data, config, [])
  end

  def trading(_position, [], _config, results), do: results
  def trading(position, [data | rest], config, results) do
    {action, price, position} =
      cond do
        create_position?(position, data, config) -> 
          create_position(position, data, config)
        add_position?(position, data, config) -> 
          add_position(position, data, config)
        close_position?(position, data, config) -> 
          close_position(position, data, config)
        stop_loss?(position, data, config) -> 
          stop_loss(position, data, config)
        true -> 
          {"short", 0, position}
      end
    
    result = %{
      date: data.date,
      tread: position.tread,
      action: action,
      price: price,
      ratio: Float.round((cast(position, data) - config.account) / config.account, 2),
    }
    
    trading(position, rest, config, results ++ [result])
  end

  def cast(%{amount: amount} = position, _data) when amount == 0, do: position.account
  def cast(position, data) do
    %{account: account, tread: tread, amount: amount, unit: unit, avg_price: avg_price} = position
    %{close: close} = data
    price = Decimal.to_float(close)

    if tread == :bull do
      account + price * unit * amount
    else
      account + (avg_price * 2 - price) * unit * amount
    end
  end

  def init_data(symbol, user) do
    stock = Markets.get_stock(symbol: symbol)
    list_dailyk = Markets.list_stock_dayk(symbol: symbol)
    list_state = Markets.list_stock_state(symbol: symbol)
    config = Accounts.get_config(user_id: user.id) |> Map.put(:lot_size, stock.lot_size)
    begin_date = Timex.shift(DateTime.utc_now(), years: -3) |> Timex.to_date()

    data = 
      Enum.zip(list_dailyk, list_state)
      |> Enum.map(fn({dailyk, state}) -> 
        dailyk_map = Map.from_struct(dailyk)
        state_map = Map.from_struct(state)
        Map.merge(dailyk_map, state_map)
      end)
      |> Enum.filter(&(Date.compare(&1.date, begin_date) == :gt))
    
    {init_position(config.account, ""), data, config}
  end

  def init_position(account, tread) do
    %{
      account: account,
      tread: tread,
      amount: 0,
      unit: 0,
      atr: 0,
      break_price: 0,
      avg_price: 0,
      add_position_price: 0,
      stop_loss_price: 0,
    }
  end

  def create_position(position, data, config) do
    %{account: account} = position
    tread = tread(data)
    break_price = break_point(data, config)
    atr = data.atr20 |> Decimal.to_float()
    unit = Common.unit(account, atr, config.atr_account_ratio, config.lot_size)
    
    opts = [
      tread: tread, 
      position: 1, 
      add_step: config.atr_add_step, 
      stop_step: config.atr_stop_step,
      atr_account_ratio: config.atr_account_ratio
    ]    
    
    position = %{
      account: account - break_price * unit,
      tread: tread,
      amount: 1,
      unit: unit,
      atr: atr,
      break_price: break_price,
      avg_price: break_price,
      add_position_price: Common.buy(break_price, atr, Keyword.put(opts, :position, 2)),
      stop_loss_price: Common.stop_loss(break_price, atr, opts)
    }

    {"create", break_price, position}
  end

  def add_position(position, _data, config) do
    %{account: account, tread: tread, amount: amount, unit: unit, atr: atr, break_price: break_price, add_position_price: add_position_price} = position
    amount = amount + 1
    opts = [
      tread: tread, 
      position: amount, 
      add_step: config.atr_add_step, 
      stop_step: config.atr_stop_step,
      atr_account_ratio: config.atr_account_ratio
    ]

    position = %{
      account: account - add_position_price * unit,
      tread: tread,
      amount: amount,
      unit: unit,
      atr: atr,
      break_price: break_price,
      avg_price: Common.buy_avg(break_price, atr, opts),
      add_position_price: Common.buy(break_price, atr, Keyword.update!(opts, :position, &(&1 + 1))),
      stop_loss_price: Common.stop_loss(break_price, atr, opts)
    }

    {"add", add_position_price, position}
  end

  def close_position(position, data, config) do
    %{account: account, tread: tread, amount: amount, unit: unit, avg_price: avg_price} = position
    price = close_point(data, config)

    account =
      if tread == :bull do
        account + price * unit * amount
      else
        account + (avg_price * 2 - price) * unit * amount
      end
      
    position = init_position(account, tread)

    {"close", price, position}
  end

  def stop_loss(position, _data, _config) do
    %{account: account, tread: tread, amount: amount, unit: unit, avg_price: avg_price, stop_loss_price: stop_loss_price} = position

    account =
      if tread == :bull do
        account + stop_loss_price * unit * amount
      else
        account + (avg_price * 2 - stop_loss_price) * unit * amount
      end

    position = init_position(account, tread)

    {"stop", stop_loss_price, position}
  end

  def create_position?(position, data, config) do
    price = break_point(data, config)
    %{lowest: lowest, highest: highest} = data

    position.amount == 0 && 
    Decimal.to_float(lowest) < price && 
    price < Decimal.to_float(highest)
  end

  def add_position?(position, data, config) do
    price = position.add_position_price
    %{lowest: lowest, highest: highest} = data

    position.amount > 0 &&
    position.amount < config.position && 
    (if position.tread == :bull, 
    do: price < Decimal.to_float(highest), 
    else: Decimal.to_float(lowest) < price)
  end

  def close_position?(position, data, config) do
    price = close_point(data, config)
    %{lowest: lowest, highest: highest} = data

    position.amount > 0 &&
    (if position.tread == :bull, 
    do: Decimal.to_float(lowest) < price, 
    else: price < Decimal.to_float(highest))
  end

  def stop_loss?(position, data, _config) do
    price = position.stop_loss_price
    %{lowest: lowest, highest: highest} = data

    position.amount > 0 &&
    (if position.tread == :bull, 
    do: Decimal.to_float(lowest) < price, 
    else: price < Decimal.to_float(highest))
  end

  def break_point(data, config) do
    %{dcu20: dcu20, dcu60: dcu60, dcl20: dcl20, dcl60: dcl60} = data

    if bull?(data) do
      case config.create_days do
        20 -> dcu20
        60 -> dcu60
      end
    else
      case config.create_days do
        20 -> dcl20
        60 -> dcl60
      end
    end
    |> Decimal.to_float()
  end

  def close_point(data, config) do
    %{dcl10: dcl10, dcl20: dcl20, dcu10: dcu10, dcu20: dcu20} = data

    if bull?(data) do
      case config.close_days do
        10 -> dcl10
        20 -> dcl20
      end
    else
      case config.close_days do
        10 -> dcu10
        20 -> dcu20
      end
    end
    |> Decimal.to_float()
  end

  def tread(data) do
    if bull?(data), do: :bull, else: :bear
  end

  def bull?(data) do
    data.ma50 > data.ma300
  end
end

File: apps/trading_system_web/assets/js/lib/backtest.js
class Backtest {
  constructor(data, config) {
    this.data = data.slice(300, data.length)
    this.config = config
    this.current_tread = this.tread(data[data.length - 1])
    this.lineData = []
    this.position = {
      tread: '',
      amount: 0,
      unit: 0,
      price: 0,
      avgPrice: 0,
      addPositionPrice: 0,
      stopLossPrice: 0,
    }
  }

  run() {
    this.data.map(x => {
      if (this.position.amount == 0 && this.isCreatePosition(x)) {
        // 建仓
        this.createPosition(x)
      }
      else if (this.position.amount > 0 && this.position.amount < this.config.position && this.isAddPosition(x)) {
        // 加仓
        this.addPosition(x)
      }
      else if (this.position.amount > 0 && this.isClosePosition(x)) {
        // 平仓
        this.closePosition(x)
      }
      else if (this.position.amount > 0 && this.isStopLoss(x)) {
        // 止损
        this.stopLoss(x)
      }
    })

    return this.lineData
  }

  // 执行建仓
  createPosition(dataItem) {
    const {date, atr} = dataItem
    const price = this.breakPoint(dataItem)

    this.position = {
      tread: this.tread(dataItem),
      amount: 1,
      unit: 0,
      atr: atr,
      breakPrice: price,
      avgPrice: this.avgPrice(dataItem, price, atr, 1),
      addPositionPrice: this.bidPrice(dataItem, price, atr, 2),
      stopLossPrice: this.stopLossPrice(dataItem, price, atr, 1),
    }

    if (this.current_tread == this.position.tread) {
      this.lineData.push({
        action: "create",
        date: date,
        price: price,
      })
    }
  }

  // 执行加仓
  addPosition(dataItem) {
    const {tread, amount, unit, atr, breakPrice} = this.position
    const current_amount = amount + 1
    
    this.position = {
      tread: tread,
      amount: current_amount,
      unit: unit,
      atr: atr,
      breakPrice: breakPrice,
      avgPrice: this.avgPrice(dataItem, breakPrice, atr, current_amount),
      addPositionPrice: this.bidPrice(dataItem, breakPrice, atr, current_amount + 1),
      stopLossPrice: this.stopLossPrice(dataItem, breakPrice, atr, current_amount),
    }
  }

  // 执行平仓
  closePosition(dataItem) {
    if (this.current_tread == this.position.tread) {
      const {date} = dataItem
      const price = this.closePoint(dataItem)

      this.lineData.push({
        action: "close",
        date: date,
        price: price
      })
    }
    
    this.position = {
      tread: '',
      amount: 0,
      unit: 0,
      price: 0,
      avgPrice: 0,
      addPositionPrice: 0,
      stopLossPrice: 0,
    }    
  }

  // 执行止损
  stopLoss(dataItem) {
    if (this.current_tread == this.position.tread) {
      const {date} = dataItem
      const price = this.position.stopLossPrice

      this.lineData.push({
        action: "stop",
        date: date,
        price: price
      })
    }
    
    this.position = {
      tread: '',
      amount: 0,
      unit: 0,
      price: 0,
      avgPrice: 0,
      addPositionPrice: 0,
      stopLossPrice: 0,
    }
  }
  
  // 买入价
  bidPrice(dataItem, price, atr, position) {
    let num1 = price
    let num2 = atr * this.config.atr_add_step * (position - 1)

    if ((this.position.tread != '' && this.position.tread == 'bull') || this.isBull(dataItem)) {
      return num1 + num2
    }
    else {
      return num1 - num2
    }
  }

  // 平均价
  avgPrice(dataItem, price, atr, position) {
    if (position == 1) {
      return price
    }
    else {
      let positionSum = 0

      for (var i = 1; i < position; i++) {
        positionSum += i
      }

      let num1 = price * position
      let num2 = atr * this.config.atr_add_step * positionSum

      if ((this.position.tread != '' && this.position.tread == 'bull') || this.isBull(dataItem)) {
        return (num1 + num2) / position
      }
      else {
        return (num1 - num2) / position
      }
    }
  }

  // 止损价
  stopLossPrice(dataItem, price, atr, position) {
    let num1 = this.avgPrice(dataItem, price, atr, position)
    let num2 = atr * (this.config.atr_stop_step / position)

    if ((this.position.tread != '' && this.position.tread == 'bull') || this.isBull(dataItem)) {
      return num1 - num2
    }
    else {
      return num1 + num2
    }
  }
  
  // 交易方向
  tread(dataItem) {
    const {ma50, ma300} = dataItem
    return ma50 > ma300 ? 'bull' : 'bear'
  }

  // 是否做多
  isBull(dataItem) {
    return this.tread(dataItem) == 'bull'
  }

  // 是否建仓
  isCreatePosition(dataItem) {
    const price = this.breakPoint(dataItem)
    const {lowest, highest} = dataItem
    
    return lowest < price && price < highest
  }

  // 是否加仓
  isAddPosition(dataItem) {
    const price = this.position.addPositionPrice
    const {lowest, highest} = dataItem
    
    if (this.position.tread == 'bull') {
      return price < highest
    }
    else {
      return lowest < price
    }
  }

  // 是否平仓
  isClosePosition(dataItem) {
    const price = this.closePoint(dataItem)
    const {lowest, highest} = dataItem

    if (this.position.tread == 'bull') {
      return lowest < price
    }
    else {
      return price < highest
    }
  }

  // 是否止损
  isStopLoss(dataItem) {
    const price = this.position.stopLossPrice
    const {lowest, highest} = dataItem
    
    if (this.position.tread == 'bull') {
      return lowest < price
    }
    else {
      return price < highest
    }
  }

  // 计算突破点
  breakPoint(dataItem) {
    const {dcu20, dcu60, dcl20, dcl60} = dataItem

    if (this.isBull(dataItem)) {
      switch (this.config.create_days) {
        case 60: 
          return dcu60
        case 20: 
        default: 
          return dcu20
      }
    }
    else {
      switch (this.config.create_days) {
        case 60: 
          return dcl60
        case 20:
        default: 
          return dcl20
      }
    }
  }

  // 计算止盈点
  closePoint(dataItem) {
    const {dcl10, dcl20, dcu10, dcu20} = dataItem
    if (this.position.tread == 'bull') {
      switch (this.config.close_days) {
        case 20:
          return dcl20
        case 10:
        default:
          return dcl10
      }
    }
    else {
      switch (this.config.close_days) {
        case 20:
          return dcu20
        case 10:
        default:
          return dcu10
      }
    }
  }
}

export default Backtest